### DISCUSSION TITLE
Submit Button

### DISCUSSION THREAD

**Discussion by cartifon** [material-components/material-web#5217] [Q&A]
Hey! 

I've been doing some tests with the web components, and looking into the button components, I saw that the `abstract class Button` is extending the `buttonBaseClass` and implementing `FormSubmitter`, and by default, the button is a 'submit'. type from `FormSubmitterType`. 

If by default, all the buttons are `submit` type, would that mean that if I am in a form and press `enter`, the form should be submitted, right? 

I am using an Angular v17 project and importing the web components there, if I have a simple button type submit, the form works just fine, but if I add the web-component button, it does not submit the form. 

I have the imports necessary for the web-component, and it shows fine in the page, but it does not submit.

this is my HTML:

```

<form [formGroup]="myForm" onSubmit="alert('SubmitForm')">
  <div>
    <label for="fullName">Full Name:</label>
    <input type="text" id="fullName" formControlName="fullName" />
  </div>

  <div>
    <label for="email">Email:</label>
    <input type="email" id="email" formControlName="email" />
  </div>

  <div>
    <label for="age">Age:</label>
    <input type="number" id="age" formControlName="age" />
  </div>

  <div>
    <md-filled-button> Submit </md-filled-button> <!--  This one does not work on 'enter' -->
    <md-filled-button type="submit" > Submit </md-filled-button> <!--  This one does not work on 'enter' either -->
    <button type="submit" >Submit</button> <!--  This one works -->
  </div>
</form>
```

Am I doing something wrong? (I hope so)

**Comment by datvm** [comment]
Not sure if  it's due to Angular framework or not but it works pretty well for me: [Lit Playground](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQgXCJAbWF0ZXJpYWwvd2ViL2FsbFwiOyAgIn0seyJuYW1lIjoiaW5kZXguaHRtbCIsImNvbnRlbnQiOiI8IURPQ1RZUEUgaHRtbD5cbjxoZWFkPlxuICA8c2NyaXB0IHR5cGU9XCJtb2R1bGVcIiBzcmM9XCIuL3NpbXBsZS1ncmVldGluZy5qc1wiPjwvc2NyaXB0PlxuPC9oZWFkPlxuPGJvZHk-XG4gIDxmb3JtPlxuICAgIDxwPiAgICAgIFxuICAgICAgPGlucHV0IG5hbWU9XCJBXCIgLz5cbiAgICA8L3A-XG4gICAgXG4gICAgPHA-ICAgICAgXG4gICAgICA8bWQtb3V0bGluZWQtdGV4dC1maWVsZCBuYW1lPVwiQlwiPjwvbWQtb3V0bGluZWQtdGV4dC1maWVsZD5cbiAgICA8L3A-XG4gICAgXG4gICAgPG1kLWZpbGxlZC1idXR0b24-U3VibWl0PC9tZC1maWxsZWQtYnV0dG9uPlxuICA8L2Zvcm0-XG4gIFxuICA8c2NyaXB0PlxuICAgIGRvY3VtZW50LnF1ZXJ5U2VsZWN0b3IoXCJmb3JtXCIpLmFkZEV2ZW50TGlzdGVuZXIoXCJzdWJtaXRcIiwgZSA9PiB7XG4gICAgICBlLnByZXZlbnREZWZhdWx0KCk7XG4gICAgICBhbGVydChcIlN1Ym1pdFwiKTtcbiAgICB9KTtcbiAgPC9zY3JpcHQ-XG48L2JvZHk-XG4ifSx7Im5hbWUiOiJwYWNrYWdlLmpzb24iLCJjb250ZW50Ijoie1xuICBcImRlcGVuZGVuY2llc1wiOiB7XG4gICAgXCJsaXRcIjogXCJeMy4wLjBcIixcbiAgICBcIkBsaXQvcmVhY3RpdmUtZWxlbWVudFwiOiBcIl4yLjAuMFwiLFxuICAgIFwibGl0LWVsZW1lbnRcIjogXCJeNC4wLjBcIixcbiAgICBcImxpdC1odG1sXCI6IFwiXjMuMC4wXCJcbiAgfVxufSIsImhpZGRlbiI6dHJ1ZX1d).

You can press Enter on the normal `input` and the form would get submitted. There is a known bug #4182 where `<md-outlined-text-field>` (and the filled version) does not submit the form on Enter though but it's not the `<md-button>`'s fault.

**Reply by cartifon** [reply]
Thank you for your answer, that is very interesting... I tried something similar to your example but with the event on the `onsubmit` and not as an event listening, and it did not work, it is here on this [Lit Playground](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQgXCJAbWF0ZXJpYWwvd2ViL2FsbFwiOyAgIn0seyJuYW1lIjoiaW5kZXguaHRtbCIsImNvbnRlbnQiOiI8IURPQ1RZUEUgaHRtbD5cbjxoZWFkPlxuICA8c2NyaXB0IHR5cGU9XCJtb2R1bGVcIiBzcmM9XCIuL3NpbXBsZS1ncmVldGluZy5qc1wiPjwvc2NyaXB0PlxuPC9oZWFkPlxuPGJvZHk-XG4gIDxmb3JtIGlkPVwibXlGb3JtXCIgb25zdWJtaXQ9XCJoYW5kbGVTdWJtaXQoZXZlbnQpXCI-XG4gICAgPGxhYmVsIGZvcj1cImZpcnN0TmFtZVwiPkZpcnN0IE5hbWU6PC9sYWJlbD5cbiAgICA8aW5wdXQgdHlwZT1cInRleHRcIiBpZD1cImZpcnN0TmFtZVwiIG5hbWU9XCJmaXJzdE5hbWVcIiByZXF1aXJlZD5cblxuICAgIDxicj5cblxuICAgIDxsYWJlbCBmb3I9XCJsYXN0TmFtZVwiPkxhc3QgTmFtZTo8L2xhYmVsPlxuICAgIDxpbnB1dCB0eXBlPVwidGV4dFwiIGlkPVwibGFzdE5hbWVcIiBuYW1lPVwibGFzdE5hbWVcIiByZXF1aXJlZD5cblxuICAgIDxicj5cblxuICAgIDxtZC1maWxsZWQtYnV0dG9uPlN1Ym1pdDwvbWQtZmlsbGVkLWJ1dHRvbj5cbiAgPC9mb3JtPlxuXG4gIDxzY3JpcHQ-XG4gICAgLy8gSmF2YVNjcmlwdCBmdW5jdGlvbiB0byBoYW5kbGUgZm9ybSBzdWJtaXNzaW9uXG4gICAgZnVuY3Rpb24gaGFuZGxlU3VibWl0KGV2ZW50KSB7XG4gICAgICBldmVudC5wcmV2ZW50RGVmYXVsdCgpOyAvLyBQcmV2ZW50IHRoZSBkZWZhdWx0IGZvcm0gc3VibWlzc2lvblxuXG4gICAgICAvLyBHZXQgdGhlIHZhbHVlcyBvZiB0aGUgaW5wdXQgZmllbGRzXG4gICAgICBjb25zdCBmaXJzdE5hbWUgPSBkb2N1bWVudC5nZXRFbGVtZW50QnlJZCgnZmlyc3ROYW1lJykudmFsdWU7XG4gICAgICBjb25zdCBsYXN0TmFtZSA9IGRvY3VtZW50LmdldEVsZW1lbnRCeUlkKCdsYXN0TmFtZScpLnZhbHVlO1xuXG4gICAgICAvLyBTaG93IGFuIGFsZXJ0IHdpdGggdGhlIHN1Ym1pdHRlZCB2YWx1ZXNcbiAgICAgIGFsZXJ0KGBGb3JtIHN1Ym1pdHRlZCFcXG5GaXJzdCBOYW1lOiAke2ZpcnN0TmFtZX1cXG5MYXN0IE5hbWU6ICR7bGFzdE5hbWV9YCk7XG4gICAgfVxuICA8L3NjcmlwdD5cbjwvZm9ybT5cbjwvYm9keT5cbiJ9LHsibmFtZSI6InBhY2thZ2UuanNvbiIsImNvbnRlbnQiOiJ7XG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcImxpdFwiOiBcIl4zLjAuMFwiLFxuICAgIFwiQGxpdC9yZWFjdGl2ZS1lbGVtZW50XCI6IFwiXjIuMC4wXCIsXG4gICAgXCJsaXQtZWxlbWVudFwiOiBcIl40LjAuMFwiLFxuICAgIFwibGl0LWh0bWxcIjogXCJeMy4wLjBcIlxuICB9XG59IiwiaGlkZGVuIjp0cnVlfV0)

**Reply by datvm** [reply]
That's interesting indeed! I found out that as soon as a 2nd `<input>` is added, this problem happens, for my example, it happens when I add another `<input>` and for your problem, if you remove one `<input>`, the Enter key works again.

**Reply by datvm** [reply]
So I noticed that [the default button](https://lit.dev/playground/#sample=examples%2Fhello-world&project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQge2h0bWwsIGNzcywgTGl0RWxlbWVudH0gZnJvbSAnbGl0JztcbmltcG9ydCB7Y3VzdG9tRWxlbWVudCwgcHJvcGVydHl9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJztcblxuQGN1c3RvbUVsZW1lbnQoJ3NpbXBsZS1ncmVldGluZycpXG5leHBvcnQgY2xhc3MgU2ltcGxlR3JlZXRpbmcgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgc3RhdGljIHN0eWxlcyA9IGNzc2BwIHsgY29sb3I6IGJsdWUgfWA7XG5cbiAgQHByb3BlcnR5KClcbiAgbmFtZSA9ICdTb21lYm9keSc7XG5cbiAgcmVuZGVyKCkge1xuICAgIHJldHVybiBodG1sYDxidXR0b24gdHlwZT1cInN1Ym1pdFwiPlN1Ym1pdDwvYnV0dG9uPmA7XG4gIH1cbn1cbiJ9LHsibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPCFET0NUWVBFIGh0bWw-XG48aGVhZD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi9zaW1wbGUtZ3JlZXRpbmcuanNcIj48L3NjcmlwdD5cbjwvaGVhZD5cbjxib2R5PlxuICA8Zm9ybT5cbiAgICA8aW5wdXQgbmFtZT1cIkFcIiAvPlxuICAgIDxpbnB1dCBuYW1lPVwiQlwiIC8-XG4gICAgXG4gICAgPGJ1dHRvbj5IZWxsbzwvYnV0dG9uPlxuICA8L2Zvcm0-XG4gIFxuICA8c2NyaXB0PlxuICAgIGNvbnN0IGJ0biA9IGRvY3VtZW50LnF1ZXJ5U2VsZWN0b3IoXCJidXR0b25cIik7XG4gICAgYnRuLmFkZEV2ZW50TGlzdGVuZXIoXCJjbGlja1wiLCBlID0-IHtcbiAgICAgIGUucHJldmVudERlZmF1bHQoKTtcbiAgICAgIGFsZXJ0KFwiQ2xpY2tlZFwiKTtcbiAgICB9KTtcbiAgICBcbiAgICBkb2N1bWVudC5xdWVyeVNlbGVjdG9yKFwiZm9ybVwiKS5hZGRFdmVudExpc3RlbmVyKFwic3VibWl0XCIsIGUgPT4ge1xuICAgICAgZS5wcmV2ZW50RGVmYXVsdCgpO1xuICAgICAgYWxlcnQoXCJTdWJtaXRcIik7XG4gICAgfSk7XG4gIDwvc2NyaXB0PlxuPC9ib2R5PlxuIn0seyJuYW1lIjoicGFja2FnZS5qc29uIiwiY29udGVudCI6IntcbiAgXCJkZXBlbmRlbmNpZXNcIjoge1xuICAgIFwibGl0XCI6IFwiXjMuMC4wXCIsXG4gICAgXCJAbGl0L3JlYWN0aXZlLWVsZW1lbnRcIjogXCJeMi4wLjBcIixcbiAgICBcImxpdC1lbGVtZW50XCI6IFwiXjQuMC4wXCIsXG4gICAgXCJsaXQtaHRtbFwiOiBcIl4zLjAuMFwiXG4gIH1cbn0iLCJoaWRkZW4iOnRydWV9XQ) receives a `click` event when pressing Enter but not the [`md-button`](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQgXCJAbWF0ZXJpYWwvd2ViL2FsbFwiOyAgIn0seyJuYW1lIjoiaW5kZXguaHRtbCIsImNvbnRlbnQiOiI8IURPQ1RZUEUgaHRtbD5cbjxoZWFkPlxuICA8c2NyaXB0IHR5cGU9XCJtb2R1bGVcIiBzcmM9XCIuL3NpbXBsZS1ncmVldGluZy5qc1wiPjwvc2NyaXB0PlxuPC9oZWFkPlxuPGJvZHk-XG4gIDxmb3JtPlxuICAgIDxwPiAgICAgIFxuICAgICAgPGlucHV0IG5hbWU9XCJBXCIgLz5cbiAgICA8L3A-XG4gICAgXG4gICAgPHA-ICAgICAgXG4gICAgICA8bWQtb3V0bGluZWQtdGV4dC1maWVsZCBuYW1lPVwiQlwiPjwvbWQtb3V0bGluZWQtdGV4dC1maWVsZD5cbiAgICA8L3A-XG4gICAgXG4gICAgPG1kLWZpbGxlZC1idXR0b24-U3VibWl0PC9tZC1maWxsZWQtYnV0dG9uPlxuICA8L2Zvcm0-XG4gIFxuICA8c2NyaXB0PlxuICAgIGNvbnN0IGJ0biA9IGRvY3VtZW50LnF1ZXJ5U2VsZWN0b3IoXCJtZC1maWxsZWQtYnV0dG9uXCIpO1xuICAgIGJ0bi5hZGRFdmVudExpc3RlbmVyKFwiY2xpY2tcIiwgZSA9PiB7XG4gICAgICBlLnByZXZlbnREZWZhdWx0KCk7XG4gICAgICBhbGVydChcIkNsaWNrXCIpO1xuICAgIH0pO1xuICAgIFxuICAgIGRvY3VtZW50LnF1ZXJ5U2VsZWN0b3IoXCJmb3JtXCIpLmFkZEV2ZW50TGlzdGVuZXIoXCJzdWJtaXRcIiwgZSA9PiB7XG4gICAgICBlLnByZXZlbnREZWZhdWx0KCk7XG4gICAgICBhbGVydChcIlN1Ym1pdFwiKTtcbiAgICB9KTtcbiAgPC9zY3JpcHQ-XG48L2JvZHk-XG4ifSx7Im5hbWUiOiJwYWNrYWdlLmpzb24iLCJjb250ZW50Ijoie1xuICBcImRlcGVuZGVuY2llc1wiOiB7XG4gICAgXCJsaXRcIjogXCJeMy4wLjBcIixcbiAgICBcIkBsaXQvcmVhY3RpdmUtZWxlbWVudFwiOiBcIl4yLjAuMFwiLFxuICAgIFwibGl0LWVsZW1lbnRcIjogXCJeNC4wLjBcIixcbiAgICBcImxpdC1odG1sXCI6IFwiXjMuMC4wXCJcbiAgfVxufSIsImhpZGRlbiI6dHJ1ZX1d).  As a result, you cannot even cancel the `<md-button>` click to prevent the form from submitting.

I guess it's either a browser bug or a bug in `<md-button>`

**Reply by asyncliz** [reply]
It's been a few years since I played with Angular forms, but I'm pretty sure the ng form stuff will prevent auto-submitting in favor of hooking into Angular form APIs.

**Reply by cartifon** [reply]
Even with just the Web Application, it doesn't submit on 'enter', on this [Lit Playground](https://lit.dev/playground/#gist=2a8349ad24ea492474929a9d7e81d2fe) you can change the buttons there if you add the button type submit and remove the other one, and then press 'enter' on the form, the form will be submitted.

**Comment by simonh1000** [comment]
I have a similar problem using Elm. My code reads
```
 Html.form
                [ class "inner"
                , onSubmit <| NewRestoSubmit elems.name
                ]
                [ mdOutlinedTextField "Restaurant name"
                    [ type_ "text"
                    , onInput NewRestoUpdateName
                    , name "resto"
                    , value elems.name
                    ]
                    []
                , mdButton [] [ text "Lookup" ]
                ]
```

and pressing enter in the text field does **not** cause a submit
